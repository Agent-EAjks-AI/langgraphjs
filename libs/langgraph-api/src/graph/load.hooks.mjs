// This module hook is used to ensure that @langchain/langgraph package
// is imported from a consistent location.
// Accepts `{ parentURL: string }` as argument when registering the hook.
const OVERRIDE_RESOLVE = [
  // Override `@langchain/langgraph` or `@langchain/langgraph/prebuilt`,
  // but not `@langchain/langgraph-sdk`
  new RegExp(`^@langchain\/langgraph(\/.+)?$`),
  new RegExp(`^@langchain\/langgraph-checkpoint(\/.+)?$`),
];

let parentURL;

export async function initialize(args) {
  parentURL = args.parentURL;
}

export async function resolve(specifier, context, nextResolve) {
  // HACK: @tailwindcss/node internally uses an ESM loader cache, which does not play nicely with `tsx`.
  //       Node.js crashes with "TypeError [ERR_INVALID_URL_SCHEME]: The URL must be of scheme file".
  //       As it already is a valid URI, we can just short-circuit the resolution and avoid `tsx`.
  if (
    specifier.includes("@tailwindcss/node/dist/esm-cache.loader.mjs") &&
    specifier.startsWith("file://")
  ) {
    return { shortCircuit: true, url: specifier, format: "module" };
  }

  if (OVERRIDE_RESOLVE.some((regex) => regex.test(specifier))) {
    return await nextResolve(specifier, { ...context, parentURL });
  }
  return nextResolve(specifier, context);
}
