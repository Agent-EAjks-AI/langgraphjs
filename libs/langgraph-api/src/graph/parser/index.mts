import { fileURLToPath } from "node:url";
import { Worker } from "node:worker_threads";
import type { JSONSchema7 } from "json-schema";
import type { Pregel } from "@langchain/langgraph";

export interface GraphSchema {
  state: JSONSchema7 | undefined;
  input: JSONSchema7 | undefined;
  output: JSONSchema7 | undefined;
  config: JSONSchema7 | undefined;
}

export interface GraphSpec {
  sourceFile: string;
  exportSymbol: string;
}

export async function getStaticGraphSchema(
  spec: GraphSpec,
  options?: { mainThread?: boolean; timeoutMs?: number },
): Promise<Record<string, GraphSchema>> {
  if (options?.mainThread) {
    const { SubgraphExtractor } = await import("./parser.mjs");
    return SubgraphExtractor.extractSchemas(
      spec.sourceFile,
      spec.exportSymbol,
      { strict: false },
    );
  }

  return await new Promise<Record<string, GraphSchema>>((resolve, reject) => {
    const worker = new Worker(
      fileURLToPath(new URL("./parser/parser.worker.mjs", import.meta.url)),
      { argv: process.argv.slice(-1) },
    );

    // Set a timeout to reject if the worker takes too long
    const timeoutId = setTimeout(() => {
      worker.terminate();
      reject(new Error("Schema extract worker timed out"));
    }, options?.timeoutMs ?? 30000);

    worker.on("message", (result) => {
      worker.terminate();
      clearTimeout(timeoutId);
      resolve(result);
    });

    worker.on("error", reject);
    worker.postMessage(spec);
  });
}

export async function getRuntimeGraphSchema(
  graph: Pregel<any, any, any, any, any>,
): Promise<GraphSchema | undefined> {
  try {
    const {
      getInputTypeSchema,
      getOutputTypeSchema,
      getUpdateTypeSchema,
      getConfigTypeSchema,
    } = await import("@langchain/langgraph/zod/schema");

    const result = {
      state: getUpdateTypeSchema(graph),
      input: getInputTypeSchema(graph),
      output: getOutputTypeSchema(graph),
      config: getConfigTypeSchema(graph),
    } as GraphSchema;

    if (Object.values(result).every((i) => i == null)) return undefined;
    return result;
  } catch {
    // ignore
  }

  return undefined;
}
