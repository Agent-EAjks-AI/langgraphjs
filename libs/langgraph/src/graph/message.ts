import {
  BaseMessage,
  BaseMessageLike,
  coerceMessageLikeToMessage,
} from "@langchain/core/messages";
import { v4 } from "uuid";
import { StateGraph } from "./state.js";
import type { LangGraphRunnableConfig } from "../pregel/runnable_types.js";
import type { StreamMessagesHandler } from "../pregel/messages.js";

export type Messages =
  | Array<BaseMessage | BaseMessageLike>
  | BaseMessage
  | BaseMessageLike;

/**
 * Prebuilt reducer that combines returned messages.
 * Can handle standard messages and special modifiers like {@link RemoveMessage}
 * instances.
 */
export function messagesStateReducer(
  left: Messages,
  right: Messages
): BaseMessage[] {
  const leftArray = Array.isArray(left) ? left : [left];
  const rightArray = Array.isArray(right) ? right : [right];
  // coerce to message
  const leftMessages = (leftArray as BaseMessageLike[]).map(
    coerceMessageLikeToMessage
  );
  const rightMessages = (rightArray as BaseMessageLike[]).map(
    coerceMessageLikeToMessage
  );
  // assign missing ids
  for (const m of leftMessages) {
    if (m.id === null || m.id === undefined) {
      m.id = v4();
      m.lc_kwargs.id = m.id;
    }
  }
  for (const m of rightMessages) {
    if (m.id === null || m.id === undefined) {
      m.id = v4();
      m.lc_kwargs.id = m.id;
    }
  }
  // merge
  const merged = [...leftMessages];
  const mergedById = new Map(merged.map((m, i) => [m.id, i]));
  const idsToRemove = new Set();
  for (const m of rightMessages) {
    const existingIdx = mergedById.get(m.id);
    if (existingIdx !== undefined) {
      if (m._getType() === "remove") {
        idsToRemove.add(m.id);
      } else {
        idsToRemove.delete(m.id);
        merged[existingIdx] = m;
      }
    } else {
      if (m._getType() === "remove") {
        throw new Error(
          `Attempting to delete a message with an ID that doesn't exist ('${m.id}')`
        );
      }
      mergedById.set(m.id, merged.length);
      merged.push(m);
    }
  }
  return merged.filter((m) => !idsToRemove.has(m.id));
}

/** @ignore */
export class MessageGraph extends StateGraph<
  BaseMessage[],
  BaseMessage[],
  Messages
> {
  constructor() {
    super({
      channels: {
        __root__: {
          reducer: messagesStateReducer,
          default: () => [],
        },
      },
    });
  }
}

export function pushMessage(
  message: BaseMessage | BaseMessageLike,
  config: LangGraphRunnableConfig,
  options?: { stateKey?: string | null }
) {
  let stateKey: string | undefined = options?.stateKey ?? "messages";
  if (options?.stateKey === null) {
    stateKey = undefined;
  }

  // coerce to message
  const validMessage = coerceMessageLikeToMessage(message);
  if (!validMessage.id) throw new Error("Message ID is required.");

  const callbacks = (() => {
    if (Array.isArray(config.callbacks)) {
      return config.callbacks;
    }

    if (typeof config.callbacks !== "undefined") {
      return config.callbacks.handlers;
    }

    return [];
  })();

  const messagesHandler = callbacks.find(
    (cb): cb is StreamMessagesHandler =>
      "name" in cb && cb.name === "StreamMessagesHandler"
  );

  if (messagesHandler) {
    const metadata = config.metadata ?? {};
    const namespace = (
      (metadata.langgraph_checkpoint_ns ?? "") as string
    ).split("|");

    messagesHandler._emit(
      [namespace, metadata],
      validMessage,
      undefined,
      false
    );
  }

  if (stateKey) {
    config.configurable?.__pregel_send?.([[stateKey, validMessage]]);
  }

  return validMessage;
}
