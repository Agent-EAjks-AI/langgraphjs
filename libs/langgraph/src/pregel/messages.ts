import {
  BaseCallbackHandler,
  HandleLLMNewTokenCallbackFields,
  NewTokenIndices,
} from "@langchain/core/callbacks/base";
import {
  AIMessageChunk,
  BaseMessage,
  isBaseMessage,
} from "@langchain/core/messages";
import { v4 } from "uuid";

import { StreamChunk } from "./loop.js";
import { Serialized } from "@langchain/core/load/serializable";
import { TAG_HIDDEN, TAG_NOSTREAM } from "../constants.js";
import { ChatGenerationChunk, LLMResult } from "@langchain/core/outputs";
import { ChainValues } from "@langchain/core/utils/types";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Meta = [string[], Record<string, any>];

function isChatGenerationChunk(x: unknown): x is ChatGenerationChunk {
  return isBaseMessage((x as ChatGenerationChunk)?.message);
}

/**
 * A callback handler that implements stream_mode=messages.
 * Collects messages from (1) chat model stream events and (2) node outputs.
 */
export class StreamMessagesHandler extends BaseCallbackHandler {
  name = "StreamMessagesHandler";

  streamFn: (streamChunk: StreamChunk) => void;

  metadatas: Record<string, Meta> = {};

  seen: Record<string, BaseMessage> = {};

  constructor(streamFn: (streamChunk: StreamChunk) => void) {
    super();
    this.streamFn = streamFn;
  }

  _emit(meta: Meta, message: BaseMessage, dedupe = false) {
    if (
      dedupe &&
      message.id !== undefined &&
      this.seen[message.id] !== undefined
    ) {
      return;
    }
    if (message.id === undefined) {
      message.id = v4();
    }
    this.seen[message.id!] = message;
    this.streamFn([meta[0], "messages", [message, meta[1]]]);
  }

  handleChatModelStart(
    _llm: Serialized,
    _messages: BaseMessage[][],
    runId: string,
    _parentRunId?: string,
    _extraParams?: Record<string, unknown>,
    tags?: string[],
    metadata?: Record<string, unknown>
  ) {
    if (
      metadata &&
      // Include legacy LangGraph SDK tag
      (!tags || !(tags.includes(TAG_NOSTREAM) && tags.includes("nostream")))
    ) {
      this.metadatas[runId] = [
        (metadata.langgraph_checkpoint_ns as string).split("NS_SEP"),
        metadata,
      ];
    }
  }

  handleLLMNewToken(
    token: string,
    _idx: NewTokenIndices,
    runId: string,
    _parentRunId?: string,
    _tags?: string[],
    fields?: HandleLLMNewTokenCallbackFields
  ) {
    const chunk = fields?.chunk;
    if (isChatGenerationChunk(chunk) && this.metadatas[runId] !== undefined) {
      this._emit(this.metadatas[runId], chunk.message);
    } else {
      this._emit(
        this.metadatas[runId],
        new AIMessageChunk({
          content: token,
        })
      );
    }
  }

  handleLLMEnd(_output: LLMResult, runId: string) {
    delete this.metadatas[runId];
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleLLMError(_err: any, runId: string) {
    delete this.metadatas[runId];
  }

  handleChainStart(
    _chain: Serialized,
    _inputs: ChainValues,
    runId: string,
    _parentRunId?: string,
    tags?: string[],
    metadata?: Record<string, unknown>,
    _runType?: string,
    runName?: string
  ) {
    console.log(runName, metadata, tags);
    if (
      metadata !== undefined &&
      runName === metadata.langgraph_node &&
      (tags === undefined || !tags.includes(TAG_HIDDEN))
    ) {
      console.log("ADDING RUN ID", runId);
      this.metadatas[runId] = [
        (metadata.langgraph_checkpoint_ns as string).split("NS_SEP"),
        metadata,
      ];
    }
  }

  handleChainEnd(outputs: ChainValues, runId: string) {
    // console.log("ENDING CHAIN");
    const metadata = this.metadatas[runId];
    console.log("METADATA", runId, this.metadatas, outputs);
    delete this.metadatas[runId];
    console.log(metadata);
    if (metadata !== undefined) {
      if (isBaseMessage(outputs)) {
        this._emit(metadata, outputs, true);
      } else if (Array.isArray(outputs)) {
        for (const value of outputs) {
          if (isBaseMessage(value)) {
            this._emit(metadata, value, true);
          }
        }
      } else if (outputs != null && typeof outputs === "object") {
        for (const value of Object.values(outputs)) {
          if (isBaseMessage(value)) {
            this._emit(metadata, value, true);
          } else if (Array.isArray(value)) {
            for (const item of value) {
              if (isBaseMessage(item)) {
                this._emit(metadata, item, true);
              }
            }
          }
        }
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleChainError(_err: any, runId: string) {
    delete this.metadatas[runId];
  }
}
