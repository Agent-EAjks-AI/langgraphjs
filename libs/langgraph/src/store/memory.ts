import { BaseStore, type V } from "./base.js";

export class MemoryStore extends BaseStore {
  private data: Map<string, Map<string, V>> = new Map();

  async list(prefixes: string[]): Promise<Record<string, Record<string, V>>> {
    const result: Record<string, Record<string, any>> = {};
    for (const prefix of prefixes) {
      if (this.data.has(prefix)) {
        result[prefix] = Object.fromEntries(this.data.get(prefix)!);
      } else {
        result[prefix] = {};
      }
    }
    return Promise.resolve(result);
  }

  async put(writes: Array<[string, string, V | null]>): Promise<void> {
    for (const [namespace, key, value] of writes) {
      if (!this.data.has(namespace)) {
        this.data.set(namespace, new Map());
      }
      const namespaceMap = this.data.get(namespace)!;
      if (value === null) {
        namespaceMap.delete(key);
      } else {
        namespaceMap.set(key, value);
      }
    }
  }
}
