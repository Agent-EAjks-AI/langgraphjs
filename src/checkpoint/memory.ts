import { RunnableConfig } from "@langchain/core/runnables";
import {
  BaseCheckpointSaver,
  Checkpoint,
  ConfigurableFieldSpec,
} from "./base.js";

export class MemorySaver extends BaseCheckpointSaver {
  storage: Record<string, Checkpoint> = {};

  get config_specs(): ConfigurableFieldSpec[] {
    return [
      {
        id: "thread_id",
        name: "Thread ID",
        annotation: null,
        description: null,
        default: null,
        isShared: true,
        dependencies: null,
      },
    ];
  }

  get(config: RunnableConfig): Checkpoint | undefined {
    return this.storage[config.configurable?.thread_id];
  }

  put(config: RunnableConfig, checkpoint: Checkpoint): void {
    this.storage[config.configurable?.thread_id] = checkpoint;
  }
}
