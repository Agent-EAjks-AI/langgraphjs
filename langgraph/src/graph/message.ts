import { BaseMessage } from "@langchain/core/messages";
import { StateGraph } from "./state.js";

type Messages = Array<BaseMessage> | BaseMessage;

export class MessageGraph extends StateGraph<BaseMessage[], Messages> {
  constructor() {
    super({
      channels: {
        __root__: [],
      },
    });
  }
}

export interface MessagesState {
  messages: BaseMessage[];
}
