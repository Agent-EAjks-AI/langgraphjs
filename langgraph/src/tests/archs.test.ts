import { it, expect } from "@jest/globals";

/*
from langchain_core.messages import HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI

from langgraph.graph import END, START, StateGraph
from langgraph.graph.message import add_messages

# useful to generate SQL query
model_low_temp = ChatOpenAI(temperature=0.1)
# useful to generate natural language outputs
model_high_temp = ChatOpenAI(temperature=0.7)

class State(TypedDict):
    # to track conversation history
    messages: Annotated[list, add_messages]
    # input
    user_query: str
    # output
    sql_query: str
    sql_explanation: str

class Input(TypedDict):
    user_query: str

class Output(TypedDict):
    sql_query: str
    sql_explanation: str

generate_prompt = SystemMessage(
    "You are a helpful data analyst, who generates SQL queries for users based on their questions."
)

def generate_sql(state: State) -> State:
    user_message = HumanMessage(state["user_query"])
    messages = [generate_prompt, *state["messages"], user_message]
    res = model_low_temp.invoke(messages)
    return {
        "sql_query": res.content,
        # update conversation history
        "messages": [user_message, res],
    }

explain_prompt = SystemMessage(
    "You are a helpful data analyst, who explains SQL queries to users."
)

def explain_sql(state: State) -> State:
    messages = [
        explain_prompt,
        # contains the user's query and the SQL query from the previous step
        *state["messages"],
    ]
    res = model_high_temp.invoke(messages)
    return {
        "sql_explanation": res.content,
        # update conversation history
        "messages": res,
    }

builder = StateGraph(State, input=Input, output=Output)
builder.add_node("generate_sql", generate_sql)
builder.add_node("explain_sql", explain_sql)
builder.add_edge(START, "generate_sql")
builder.add_edge("generate_sql", "explain_sql")
builder.add_edge("explain_sql", END)

graph = builder.compile()
*/
import {
  AIMessage,
  SystemMessage,
  HumanMessage,
} from "@langchain/core/messages";
import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { DocumentInterface, Document } from "@langchain/core/documents";
import { DuckDuckGoSearch } from "@langchain/community/tools/duckduckgo_search";
import { Calculator } from "@langchain/community/tools/calculator";
import { ToolNode, toolsCondition } from "../prebuilt/index.js";
import {
  StateGraph,
  Annotation,
  messagesStateReducer,
  START,
  END,
} from "../web.js";

it("chain arch", () => {
  // useful to generate SQL query
  const modelLowTemp = new ChatOpenAI({ temperature: 0.1 });
  // useful to generate natural language outputs
  const modelHighTemp = new ChatOpenAI({ temperature: 0.7 });

  const annotation = Annotation.Root({
    messages: Annotation({ reducer: messagesStateReducer, default: () => [] }),
    user_query: Annotation<string>(),
    sql_query: Annotation<string>(),
    sql_explanation: Annotation<string>(),
  });

  const generatePrompt = new SystemMessage(
    "You are a helpful data analyst, who generates SQL queries for users based on their questions."
  );

  async function generateSql(state: typeof annotation.State) {
    const userMessage = new HumanMessage(state.user_query);
    const messages = [generatePrompt, ...state.messages, userMessage];
    const res = await modelLowTemp.invoke(messages);
    return {
      sql_query: res.content as string,
      // update conversation history
      messages: [userMessage, res],
    };
  }

  const explainPrompt = new SystemMessage(
    "You are a helpful data analyst, who explains SQL queries to users."
  );

  async function explainSql(state: typeof annotation.State) {
    const messages = [explainPrompt, ...state.messages];
    const res = await modelHighTemp.invoke(messages);
    return {
      sql_explanation: res.content as string,
      // update conversation history
      messages: res,
    };
  }

  const builder = new StateGraph(annotation)
    .addNode("generate_sql", generateSql)
    .addNode("explain_sql", explainSql)
    .addEdge(START, "generate_sql")
    .addEdge("generate_sql", "explain_sql")
    .addEdge("explain_sql", END);

  const graph = builder.compile();

  expect(graph.getGraph().drawMermaid()).toMatchSnapshot();
});

/*
from typing import Annotated, Literal, TypedDict

from langchain_core.messages import HumanMessage, SystemMessage
from langchain_core.vectorstores.in_memory import InMemoryVectorStore
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

from langgraph.graph import END, START, StateGraph
from langgraph.graph.message import add_messages

embeddings = OpenAIEmbeddings()
# useful to generate SQL query
model_low_temp = ChatOpenAI(temperature=0.1)
# useful to generate natural language outputs
model_high_temp = ChatOpenAI(temperature=0.7)

class State(TypedDict):
    # to track conversation history
    messages: Annotated[list, add_messages]
    # input
    user_query: str
    # output
    domain: Literal["records", "insurance"]
    documents: str
    answer: str

class Input(TypedDict):
    user_query: str

class Output(TypedDict):
    documents: str
    answer: str

# refer to chapter 2 on how to fill a vector store with documents
medical_records_store = InMemoryVectorStore.from_documents([], embeddings)
medical_records_retriever = medical_records_store.as_retriever()

insurance_faqs_store = InMemoryVectorStore.from_documents([], embeddings)
insurance_faqs_retriever = insurance_faqs_store.as_retriever()

router_prompt = SystemMessage(
    """You need to decide which domain to route the user query to. You have two domains to choose from:
- records: contains medical records of the patient, such as diagnosis, treatment, and prescriptions.
- insurance: contains frequently asked questions about insurance policies, claims, and coverage.

Output only the domain name."""
)

def router_node(state: State) -> State:
    user_message = HumanMessage(state["user_query"])
    messages = [router_prompt, *state["messages"], user_message]
    res = model_low_temp.invoke(messages)
    return {
        "domain": res.content,
        # update conversation history
        "messages": [user_message, res],
    }

def pick_retriever(
    state: State,
) -> Literal["retrieve_medical_records", "retrieve_insurance_faqs"]:
    if state["domain"] == "records":
        return "retrieve_medical_records"
    else:
        return "retrieve_insurance_faqs"

def retrieve_medical_records(state: State) -> State:
    documents = medical_records_retriever.invoke(state["user_query"])
    return {
        "documents": documents,
    }

def retrieve_insurance_faqs(state: State) -> State:
    documents = insurance_faqs_retriever.invoke(state["user_query"])
    return {
        "documents": documents,
    }

medical_records_prompt = SystemMessage(
    "You are a helpful medical chatbot, who answers questions based on the patient's medical records, such as diagnosis, treatment, and prescriptions."
)

insurance_faqs_prompt = SystemMessage(
    "You are a helpful medical insurance chatbot, who answers frequently asked questions about insurance policies, claims, and coverage."
)

def generate_answer(state: State) -> State:
    if state["domain"] == "records":
        prompt = medical_records_prompt
    else:
        prompt = insurance_faqs_prompt
    messages = [
        prompt,
        *state["messages"],
        HumanMessage(f"Documents: {state["documents"]}"),
    ]
    res = model_high_temp.invoke(messages)
    return {
        "answer": res.content,
        # update conversation history
        "messages": res,
    }

builder = StateGraph(State, input=Input, output=Output)
builder.add_node("router", router_node)
builder.add_node("retrieve_medical_records", retrieve_medical_records)
builder.add_node("retrieve_insurance_faqs", retrieve_insurance_faqs)
builder.add_node("generate_answer", generate_answer)
builder.add_edge(START, "router")
builder.add_conditional_edges("router", pick_retriever)
builder.add_edge("retrieve_medical_records", "generate_answer")
builder.add_edge("retrieve_insurance_faqs", "generate_answer")
builder.add_edge("generate_answer", END)

graph = builder.compile()

assert graph.get_graph().draw_mermaid() == snapshot
*/
it("router arch", async () => {
  const embeddings = new OpenAIEmbeddings();
  // useful to generate SQL query
  const modelLowTemp = new ChatOpenAI({ temperature: 0.1 });
  // useful to generate natural language outputs
  const modelHighTemp = new ChatOpenAI({ temperature: 0.7 });

  const annotation = Annotation.Root({
    messages: Annotation({ reducer: messagesStateReducer, default: () => [] }),
    user_query: Annotation<string>,
    domain: Annotation<"records" | "insurance">,
    documents: Annotation<DocumentInterface[]>,
    answer: Annotation<string>,
  });

  // refer to chapter 2 on how to fill a vector store with documents
  const medicalRecordsStore = await MemoryVectorStore.fromDocuments(
    [],
    embeddings
  );
  const medicalRecordsRetriever = medicalRecordsStore.asRetriever();

  const insuranceFaqsStore = await MemoryVectorStore.fromDocuments(
    [],
    embeddings
  );
  const insuranceFaqsRetriever = insuranceFaqsStore.asRetriever();

  const routerPrompt = new SystemMessage(
    `You need to decide which domain to route the user query to. You have two domains to choose from:
- records: contains medical records of the patient, such as diagnosis, treatment, and prescriptions.
- insurance: contains frequently asked questions about insurance policies, claims, and coverage.

Output only the domain name.`
  );

  async function routerNode(state: typeof annotation.State) {
    const userMessage = new HumanMessage(state.user_query);
    const messages = [routerPrompt, ...state.messages, userMessage];
    const res = await modelLowTemp.invoke(messages);
    return {
      domain: res.content as "records" | "insurance",
      // update conversation history
      messages: [userMessage, res],
    };
  }

  function pickRetriever(state: typeof annotation.State) {
    if (state.domain === "records") {
      return "retrieve_medical_records";
    } else {
      return "retrieve_insurance_faqs";
    }
  }

  async function retrieveMedicalRecords(state: typeof annotation.State) {
    const documents = await medicalRecordsRetriever.invoke(state.user_query);
    return {
      documents,
    };
  }

  async function retrieveInsuranceFaqs(state: typeof annotation.State) {
    const documents = await insuranceFaqsRetriever.invoke(state.user_query);
    return {
      documents,
    };
  }

  const medicalRecordsPrompt = new SystemMessage(
    "You are a helpful medical chatbot, who answers questions based on the patient's medical records, such as diagnosis, treatment, and prescriptions."
  );

  const insuranceFaqsPrompt = new SystemMessage(
    "You are a helpful medical insurance chatbot, who answers frequently asked questions about insurance policies, claims, and coverage."
  );

  async function generateAnswer(state: typeof annotation.State) {
    const prompt =
      state.domain === "records" ? medicalRecordsPrompt : insuranceFaqsPrompt;
    const messages = [
      prompt,
      ...state.messages,
      new HumanMessage(`Documents: ${state.documents}`),
    ];
    const res = await modelHighTemp.invoke(messages);
    return {
      answer: res.content as string,
      // update conversation history
      messages: res,
    };
  }

  const builder = new StateGraph(annotation)
    .addNode("router", routerNode)
    .addNode("retrieve_medical_records", retrieveMedicalRecords)
    .addNode("retrieve_insurance_faqs", retrieveInsuranceFaqs)
    .addNode("generate_answer", generateAnswer)
    .addEdge(START, "router")
    .addConditionalEdges("router", pickRetriever)
    .addEdge("retrieve_medical_records", "generate_answer")
    .addEdge("retrieve_insurance_faqs", "generate_answer")
    .addEdge("generate_answer", END);

  const graph = builder.compile();

  expect(graph.getGraph().drawMermaid()).toMatchSnapshot();
});

/*
from typing import Annotated, TypedDict

from langchain_community.tools import DuckDuckGoSearchRun
from langchain_openai import ChatOpenAI

from langgraph.graph import START, StateGraph
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

search = DuckDuckGoSearchRun()
tools = [search]
model = ChatOpenAI(temperature=0.1).bind_tools(tools)

class State(TypedDict):
    messages: Annotated[list, add_messages]

def model_node(state: State) -> State:
    res = model.invoke(state["messages"])
    return {"messages": res}

builder = StateGraph(State)
builder.add_node("model", model_node)
builder.add_node("tools", ToolNode(tools))
builder.add_edge(START, "model")
builder.add_conditional_edges("model", tools_condition)
builder.add_edge("tools", "model")

graph = builder.compile()

assert graph.get_graph().draw_mermaid() == snapshot
*/

it("agent arch", () => {
  const search = new DuckDuckGoSearch();
  const calculator = new Calculator();
  const tools = [search, calculator];
  const model = new ChatOpenAI({ temperature: 0.1 }).bindTools(tools);

  const annotation = Annotation.Root({
    messages: Annotation({ reducer: messagesStateReducer, default: () => [] }),
  });

  async function modelNode(state: typeof annotation.State) {
    const res = await model.invoke(state.messages);
    return { messages: res };
  }

  const builder = new StateGraph(annotation)
    .addNode("model", modelNode)
    .addNode("tools", new ToolNode<typeof annotation.State>(tools))
    .addEdge(START, "model")
    .addConditionalEdges("model", toolsCondition)
    .addEdge("tools", "model");

  const graph = builder.compile();
});

it("agent arch always tool", () => {
  const search = new DuckDuckGoSearch();
  const calculator = new Calculator();
  const tools = [search, calculator];
  const model = new ChatOpenAI({ temperature: 0.1 }).bindTools(tools);

  const annotation = Annotation.Root({
    messages: Annotation({ reducer: messagesStateReducer, default: () => [] }),
  });

  async function firstModelNode(state: typeof annotation.State) {
    const query = state.messages[state.messages.length - 1].content;
    const searchToolCall = {
      name: "duckduckgo_search",
      args: { query },
      id: Math.random().toString(),
    };
    return {
      messages: [new AIMessage({ content: "", tool_calls: [searchToolCall] })],
    };
  }

  async function modelNode(state: typeof annotation.State) {
    const res = await model.invoke(state.messages);
    return { messages: res };
  }

  const builder = new StateGraph(annotation)
    .addNode("first_model", firstModelNode)
    .addNode("model", modelNode)
    .addNode("tools", new ToolNode<typeof annotation.State>(tools))
    .addEdge(START, "first_model")
    .addEdge("first_model", "tools")
    .addEdge("tools", "model")
    .addConditionalEdges("model", toolsCondition);

  const graph = builder.compile();
});

/*
import ast
from typing import Annotated, TypedDict

from langchain_community.tools import DuckDuckGoSearchRun
from langchain_core.documents import Document
from langchain_core.messages import HumanMessage
from langchain_core.tools import tool
from langchain_core.vectorstores.in_memory import InMemoryVectorStore
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

from langgraph.graph import START, StateGraph
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

@tool
def calculator(query: str) -> str:
    """A simple calculator tool. Input should be a mathematical expression."""
    return ast.literal_eval(query)

search = DuckDuckGoSearchRun()
tools = [search, calculator]
embeddings = OpenAIEmbeddings()
model = ChatOpenAI(temperature=0.1)
tools_retriever = InMemoryVectorStore.from_documents(
    [Document(tool.description, metadata={"name": tool.name}) for tool in tools],
    embeddings,
).as_retriever()

class State(TypedDict):
    messages: Annotated[list, add_messages]
    selected_tools: list[str]

def model_node(state: State) -> State:
    selected_tools = [
        tool for tool in tools if tool.name in state["selected_tools"]
    ]
    res = model.bind_tools(selected_tools).invoke(state["messages"])
    return {"messages": res}

def select_tools(state: State) -> State:
    query = state["messages"][-1].content
    tool_docs = tools_retriever.invoke(query)
    return {"selected_tools": [doc.metadata["name"] for doc in tool_docs]}

builder = StateGraph(State)
builder.add_node("select_tools", select_tools)
builder.add_node("model", model_node)
builder.add_node("tools", ToolNode(tools))
builder.add_edge(START, "select_tools")
builder.add_edge("select_tools", "model")
builder.add_conditional_edges("model", tools_condition)
builder.add_edge("tools", "model")

graph = builder.compile()
*/
it("agent arch with tools", async () => {
  const search = new DuckDuckGoSearch();
  const calculator = new Calculator();
  const tools = [search, calculator];
  const embeddings = new OpenAIEmbeddings();
  const model = new ChatOpenAI({ temperature: 0.1 });
  const toolsStore = await MemoryVectorStore.fromDocuments(
    tools.map(
      (tool) =>
        new Document({
          pageContent: tool.description,
          metadata: { name: tool.constructor.name },
        })
    ),
    embeddings
  );
  const toolsRetriever = toolsStore.asRetriever();

  const annotation = Annotation.Root({
    messages: Annotation({ reducer: messagesStateReducer, default: () => [] }),
    selected_tools: Annotation<string[]>(),
  });

  async function modelNode(state: typeof annotation.State) {
    const selectedTools = tools.filter((tool) =>
      state.selected_tools.includes(tool.constructor.name)
    );
    const res = await model.bindTools(selectedTools).invoke(state.messages);
    return { messages: res };
  }

  async function selectTools(state: typeof annotation.State) {
    const query = state.messages[state.messages.length - 1].content;
    const toolDocs = await toolsRetriever.invoke(query as string);
    return {
      selected_tools: toolDocs.map((doc) => doc.metadata.name),
    };
  }

  const builder = new StateGraph(annotation)
    .addNode("select_tools", selectTools)
    .addNode("model", modelNode)
    .addNode("tools", new ToolNode<typeof annotation.State>(tools))
    .addEdge(START, "select_tools")
    .addEdge("select_tools", "model")
    .addConditionalEdges("model", toolsCondition)
    .addEdge("tools", "model");

  const graph = builder.compile();
});
