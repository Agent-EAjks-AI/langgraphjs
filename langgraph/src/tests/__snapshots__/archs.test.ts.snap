// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chain arch 1`] = `
"%%{init: {'flowchart': {'curve': 'linear'}}}%%
graph TD;
	__start__[__start__]:::startclass;
	__end__[__end__]:::endclass;
	generate_sql([generate_sql]):::otherclass;
	explain_sql([explain_sql]):::otherclass;
	__start__ --> generate_sql;
	generate_sql --> explain_sql;
	explain_sql --> __end__;
	classDef startclass fill:#ffdfba;
	classDef endclass fill:#baffc9;
	classDef otherclass fill:#fad7de;
"
`;
